Enhance RecordFunctionFast input args and use input args in triton_heuristics.py

sraikund16 opened the pull request:
Summary: Now that we can input shapes as input args for RecordFunctionFast, let's add that to the triton heuristics. Also, lets add the ability to pass in a tuple into the RecordFunctionFast constructor.

Test Plan:
Ran both the _inductor/test_profile.py and profiler/test_profiler.py unit tests. Also added tuple based unit test to profiler/test_profiler.py

Ran record_function_fast.py from the following branch
https://github.com/pytorch/pytorch/compare/sraikund/record_funct_test?expand=1

No shape or args: tests function fast with no args and profile without record_shapes
With shape tests: tests function fast with args and profile with record_shapes true
Args no shape: tests function fast with args inputted but record_shapes set to false
Args shape tuple: tests function fast with args inputted in form of tuple and record_shapes true


Stdout:

No shape or args:: 1.8491458892822266 us
With shape:: 2.211381196975708 us
Args no shape:: 1.9212646484375 us
With shape tuple:: 2.245788335800171 us

Differential Revision: D55809967




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

davidberard98 reviewed (approved):
pending tests, LGTM!

peterbell10 commented on torch/csrc/profiler/python/init.cpp:
```diff
@@ -210,8 +213,17 @@ PyObject* RecordFunctionFast_enter(PyObject* selfGeneric, PyObject* unused) {
       std::vector<at::IValue> ivalues;
-      for (int i = 0; i < PyList_Size(self->inputValues); i++) {
-        PyObject* item = PyList_GetItem(self->inputValues, i);
+      std::function<Py_ssize_t(PyObject*)> getSize;
+      std::function<PyObject*(PyObject*, Py_ssize_t)> getItem;
+      if (PyTuple_Check(self->inputValues)) {
+        getSize = PyTuple_Size;
+        getItem = PyTuple_GetItem;
+      } else {
+        getSize = PyList_Size;
+        getItem = PyList_GetItem;
+      }
+      for (int i = 0; i < getSize(self->inputValues); i++) {
+        PyObject* item = getItem(self->inputValues, i);
```
How about:
```suggestion
      THPObjectPtr input_fast(PySequence_Fast(self->inputValues, "input must be a sequence"));
      PyObject ** input_items = PySequence_Fast_ITEMS(input_fast);
      for (int i = 0; i < PySequence_Fast_GET_SIZE(input_fast); i++) {
        PyObject* item = input_items[i];
```

sraikund16 replied:
Thanks for the suggestion! Let me try this out.

sraikund16 committed:
Enhance RecordFunctionFast input args and use input args in triton_heuristics.py (#123459)

Summary:
Now that we can input shapes as input args for RecordFunctionFast, let's add that to the triton heuristics. Also, lets add the ability to pass in a tuple into the RecordFunctionFast constructor.


Test Plan:
Ran both the _inductor/test_profile.py and profiler/test_profiler.py unit tests. Also added tuple based unit test to profiler/test_profiler.py

Ran record_function_fast.py from the following branch
https://github.com/pytorch/pytorch/compare/sraikund/record_funct_test?expand=1

No shape or args: tests function fast with no args and profile without record_shapes
With shape tests: tests function fast with args and profile with record_shapes true
Args no shape: tests function fast with args inputted but record_shapes set to false
Args shape tuple: tests function fast with args inputted in form of tuple and record_shapes true


Stdout:

No shape or args:: 1.8491458892822266 us
With shape:: 2.211381196975708 us
Args no shape:: 1.9212646484375 us
With shape tuple:: 2.245788335800171 us





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: D55809967

Pulled By: sraikund16
